### PR TITLE
[Doc] Update Python test doc

### PR DESCRIPTION
**Pull request by ljcc0930**

### CODE REVIEW

File: docs/lang/articles/contribution/write_test.md [hunk 1]
```diff
@@ -62,7 +62,7 @@ The line `ti.init(arch=ti.cpu)` in the test above means that it will only test o
 import taichi as ti
 
 # will test against both CPU and CUDA backends
-@ti.test(ti.cpu, ti.cuda)
+@ti.test(arch = [ti.cpu, ti.cuda])
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/write_test.md:65]:
```suggestion
@ti.test(arch=[ti.cpu, ti.cuda])
```

File: docs/lang/articles/contribution/write_test.md [hunk 2]
```diff
@@ -30,7 +30,7 @@ def test_log10():
     pass
 ```
 
-Add some simple code that makes use of `ti.log10` to ensure it works
+Add some tests that makes use of `ti.log10` to ensure it works
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/write_test.md:33]:
```suggestion
Add some tests that make use of `ti.log10` to ensure it works
```

File: docs/lang/articles/contribution/write_test.md [hunk 3]
```diff
@@ -75,6 +75,8 @@ def test_log10():
     assert r[None] == 2
 ```
 
+In this case, python test will execute this function on `ti.cpu` and `ti.cuda` backend separately.
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/write_test.md:78]:
```suggestion
In this case, Taichi will execute this test case on both `ti.cpu` and `ti.cuda` backends.
```

File: docs/lang/articles/contribution/write_test.md [hunk 4]
```diff
@@ -98,7 +100,7 @@ def test_log10():
 ## Using `ti.approx` for comparison with tolerance
 
 Sometimes the precision of math operations could be relatively low on certain backends such as OpenGL,
-e.g. `ti.log10(100)` may return `2.001` or `1.999` in this case.
+e.g. `ti.log10(100)` may return `2.000001` or `1.999999` in this case.
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/write_test.md:103]:
```suggestion
e.g., `ti.log10(100)` may return `2.000001` or `1.999999` in this case.
```

File: docs/lang/articles/contribution/write_test.md [hunk 5]
```diff
@@ -241,3 +243,46 @@ a specific feature:
 def test_sparse_field():
     # ... (some tests that requires sparse feature which is not supported by OpenGL)
 ```
+
+## Requiring extension in tests
+
+If the test is depend on some extensions, you could add argument `require` in `@ti.test` decorator.
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/write_test.md:249]:
```suggestion
If a test case depends on some extensions, you should add an argument `require` in the `@ti.test` decorator.
```

File: docs/lang/articles/contribution/write_test.md [hunk 6]
```diff
@@ -241,3 +243,46 @@ a specific feature:
...
+    def count() -> int:
+        tot = 0
+        for I in ti.grouped(block):
+            tot += 1
+        return tot
+
+    assert count() == 1
+```
+
+Now, Taichi is supporting following extensions:
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/write_test.md:274]:
```suggestion
Now, Taichi supports the following extensions:
```

File: docs/lang/articles/contribution/write_test.md [hunk 7]
```diff
@@ -241,3 +243,46 @@ a specific feature:
...
+
+| Name          | Extension usage                                               |
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/write_test.md:276]:
```suggestion
| Name          | Extension details                                             |
```

File: docs/lang/articles/contribution/write_test.md [hunk 8]
```diff
@@ -241,3 +243,46 @@ a specific feature:
...
+| ------------- | ------------------------------------------------------------- |
+| sparse        | Sparse data structures                                        |
+| async_mode    | Asynchronous execution mode                                   |
+| quant_basic   | Basic operations in quantization                              |
+| quant         | Quantization                                                  |
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/write_test.md:281]:
```suggestion
| quant         | Full quantization functionalities                             |
```

Comment (yuanming-hu) [inline docs/lang/articles/contribution/write_test.md:281]:
```suggestion
| quant         | Basic quantization functionalities                            |
```

File: docs/lang/articles/contribution/write_test.md [hunk 9]
```diff
@@ -241,3 +243,46 @@ a specific feature:
...
+| data64        | 64-bit data (Metal doesn't support 64-bit data buffers yet)   |
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/write_test.md:282]:
```suggestion
| data64        | 64-bit data and arithmetics                                   |
```

File: docs/lang/articles/contribution/write_test.md [hunk 10]
```diff
@@ -241,3 +243,46 @@ a specific feature:
...
+| adstack       | For keeping the history of mutable local variables            |
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/write_test.md:283]:
```suggestion
| adstack       | For keeping the history of mutable local variables in autodiff|
```

File: docs/lang/articles/contribution/write_test.md [hunk 11]
```diff
@@ -241,3 +243,46 @@ a specific feature:
...
+| bls           | Block-local storage                                           |
+| assertion     | Run-time asserts in Taichi kernels                            |
+| extfunc       | Invoke external functions or backend source                   |
+| packed        | Packed mode, Shape will not be padded to a power of two       |
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/write_test.md:287]:
```suggestion
| packed        | Packed mode: shapes will not be padded to powers of two       |
```

File: docs/lang/articles/contribution/write_test.md [hunk 12]
```diff
@@ -241,3 +243,46 @@ a specific feature:
...
+| dynamic_index | Dynamic index support for both global and local tensors       |
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/write_test.md:288]:
```suggestion
| dynamic_index | Dynamic index support for matrices and vectors                |
```

File: docs/lang/articles/contribution/write_test.md [hunk 13]
```diff
@@ -241,3 +243,46 @@ a specific feature:
...
+| ------------- | ------------------------------------------------------------- |
+| sparse        | Sparse data structures                                        |
+| async_mode    | Asynchronous execution mode                                   |
+| quant_basic   | Basic operations in quantization                              |
+| quant         | Quantization                                                  |
+| data64        | 64-bit data (Metal doesn't support 64-bit data buffers yet)   |
+| adstack       | For keeping the history of mutable local variables            |
+| bls           | Block-local storage                                           |
+| assertion     | Run-time asserts in Taichi kernels                            |
+| extfunc       | Invoke external functions or backend source                   |
```
Comment (yuanming-hu) [inline docs/lang/articles/contribution/write_test.md:286]:
```suggestion
| extfunc       | Support inserting external functions calls or backend source  |
```

**Review by yuanming-hu — COMMENTED** [review]
Cool! Some nits.

File: docs/lang/articles/contribution/write_test.md [hunk 14]
```diff
@@ -75,6 +75,8 @@ def test_log10():
     assert r[None] == 2
 ```
 
+In this case, Taichi will execute this test case on both `ti.cpu` and `ti.cuda` backends.
+
 And you may test against **all backends** by simply not specifying the
```
Comment (qiao-bo) [inline docs/lang/articles/contribution/write_test.md:80]: nit: ...all (available?) backends... or backends (depending on the system)... ;/

File: docs/lang/articles/contribution/write_test.md [hunk 15]
```diff
@@ -98,7 +100,7 @@ def test_log10():
 ## Using `ti.approx` for comparison with tolerance
 
 Sometimes the precision of math operations could be relatively low on certain backends such as OpenGL,
```
Comment (qiao-bo) [inline docs/lang/articles/contribution/write_test.md:102]: nit: ...operations could be limited...? ;/

File: docs/lang/articles/contribution/write_test.md [hunk 16]
```diff
@@ -241,3 +243,46 @@ a specific feature:
...
+| ------------- | ------------------------------------------------------------- |
+| sparse        | Sparse data structures                                        |
+| async_mode    | Asynchronous execution mode                                   |
+| quant_basic   | Basic operations in quantization                              |
+| quant         | Full quantization functionalities                             |
+| data64        | 64-bit data and arithmetics                                   |
+| adstack       | For keeping the history of mutable local variables in autodiff|
+| bls           | Block-local storage                                           |
+| assertion     | Run-time asserts in Taichi kernels                            |
+| extfunc       | Support inserting external functions calls or backend source  |
```
Comment (qiao-bo) [inline docs/lang/articles/contribution/write_test.md:286]: ... external function calls or...

**Review by qiao-bo — COMMENTED** [review]
Thanks! Just some tiny nits ;/

**Comment by ljcc0930** [comment]
> Thanks! Just some tiny nits ;/

Thanks! Will fix them soon!

**Review by qiao-bo — APPROVED** [review]

**Pull request merged**